decoder/src/log/format.rs

```diff
@@ -0,0 +1,115 @@
+use nom::branch::alt;
+use nom::bytes::complete::{take, take_till1};
+use nom::character::complete::char;
+use nom::combinator::{map, map_res};
+use nom::error::{FromExternalError, ParseError};
+use nom::multi::many0;
+use nom::sequence::delimited;
+use nom::IResult;
+use nom::Parser;
+
+#[derive(Debug, PartialEq, Clone)]
+#[non_exhaustive]
+pub enum LogSegment {
+    String(String),
+    Timestamp,
+    FileName,
+    FilePath,
+    ModulePath,
+    LineNumber,
+    LogLevel,
+    Log,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct InvalidArgument;
+
+fn parse_argument<'a, E>(input: &'a str) -> IResult<&'a str, LogSegment, E>
+where
+    E: ParseError<&'a str> + FromExternalError<&'a str, InvalidArgument>,
+{
+    let parse_enclosed = delimited(char('{'), take(1u32), char('}'));
+    let mut parse_type = map_res(parse_enclosed, move |s| match s {
+        "t" => Ok(LogSegment::Timestamp),
+        "f" => Ok(LogSegment::FileName),
+        "F" => Ok(LogSegment::FilePath),
+        "m" => Ok(LogSegment::ModulePath),
+        "l" => Ok(LogSegment::LineNumber),
+        "L" => Ok(LogSegment::LogLevel),
+        "s" => Ok(LogSegment::Log),
+        _ => Err(InvalidArgument),
+    });
+
+    parse_type.parse(input)
+}
+
+fn parse_string_segment<'a, E>(input: &'a str) -> IResult<&'a str, LogSegment, E>
+where
+    E: ParseError<&'a str>,
+{
```

Comment:
Could the error type just be `()`? I don't see a call-site where you are using a different concrete type. Applies to this and the other functions.

```suggestion
fn parse_string_segment<'a, E>(input: &'a str) -> IResult<&'a str, LogSegment, ()> {
```